### PR TITLE
Exposing prepare_for_model for both slow & fast tokenizers

### LINKED ISSUE #5377: New tokenizer code in transformer 3.0.0 is creating error with old code
# 🐛 Bug

## Information

Model I am using (Bert, XLNet ...): BeRT and GPT2 for Poly-encoder implementation

Language I am using the model on (English, Chinese ...): English

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)
Ubuntu V2.0 corpus dataset, implementing Poly-encoder pipeline, everything was done, was re-training the model again to verify the results of the first training process.

## To reproduce

Happens when using some_tokenizer_fromHF.encode_plus(), below is the eval script to test on custom input text, is exactly same as the one reading from dataset during training(simplified for eval)
```
from transformers import BertTokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
c_text="what is your name. I am Gloid"
context=tokenizer.encode_plus(c_text,
                                      text_pair=None,
                                      add_special_tokens=True,
                                      max_length=max_len,
                                      pad_to_max_length=False)
texts=["Obama", "Trump", "Eminem", "slender man", "Pewdiepie"]
for text in texts:
      tokenized_dict = tokenizer.encode_plus(text,
                                                  text_pair=None,
                                                  add_special_tokens=True,
                                                  max_length=max_len,
                                                  pad_to_max_length=True)
```
The error is - 
"Truncation was not explicitely activated but `max_length` is provided a specific value, please use `truncation=True` to explicitely truncate examples to max length. Defaulting to 'only_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you may want to check this is the right behavior."
Repeated a total of 6 times, i.e.,  for every sequence passed into encode_plus

## Expected behavior

Not to give this error, and return the input ids, segment ids, and input masks
The issue is completely identical to the closed issue - https://github.com/huggingface/transformers/issues/5155

## Environment info

     
- `transformers` version: 3.0.0
- Platform: Linux-4.19.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.6.9
- PyTorch version (GPU?): 1.5.1+cu101 (True)
- Tensorflow version (GPU?): 2.2.0 (True)
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### LINKED ISSUE #5447: Where did "prepare_for_model" go?  What is the replacement?
I'm working with already numericalized data (e.g., where the text has been converted to ids via `tokenizer.tokenize()`) and was using `prepare_for_model` to build the appropriate input dictionary ... ***but*** that method is gone in 3.0.

So ... what should I use/do now?

Thanks

### LINKED ISSUE #5460: BERT Huggingface trainer api: ValueError: expected sequence of length 128 at dim 1 (got 314)
# ❓ Questions & Help

## Details

I'm using new trainer api in HuggingFace Transformers to train on a GLUE task (QQP). This error shows up during training.

This is the example I'm using https://colab.research.google.com/github/huggingface/blog/blob/master/notebooks/trainer/01_text_classification.ipynb

Only change I made is the task. In the notebook GLUE task is MNLI which I changed to QQP. While original MNLI task runs without errors, QQP task fails.

ValueError: expected sequence of length 128 at dim 1 (got 314)

Please check my Stackoverflow question for more details
**https://stackoverflow.com/questions/62675482/bert-huggingface-trainer-api-valueerror-expected-sequence-of-length-128-at-dim**:

### PR DESCRIPTION
**Pull request by LysandreJik**
With version v3.0.0, two breaking changes that could have been avoided have been introduced. After discussion with @n1to and @thomwolf, this PR aims to revert these changes, by implementing two changes:

- The `prepare_for_model` method for both slow and tokenizers is now publicly exposed (it was only the case for the slow tokenizers before v3.0.0)
- The truncation methods now default to `longest_first` instead of `first_only` .

This PR adds two tests, for both Python and Rust tokenizers:

- Assert that `tokenizer.prepare_for_model(tokenizer.encode(x)) == tokenizer.encode_plus(x)`
- Assert that the output of `prepare_for_model` for rust and python tokenizers is equal.

closes https://github.com/huggingface/transformers/issues/5377
closes https://github.com/huggingface/transformers/issues/5447
closes https://github.com/huggingface/transformers/issues/5460

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -1960,6 +1960,219 @@ def pad(
...
+                FutureWarning
+            )
+            return_length = return_lengths
+
+        if pad_to_max_length:
+            warnings.warn(
+                "The PreTrainedTokenizerBase.prepare_for_model `pad_to_max_length` parameter is deprecated."
+                "Please use `padding_strategy` instead.",
+                FutureWarning
+            )
```
Comment (LysandreJik) [inline src/transformers/tokenization_utils_base.py:2024]: Maybe hiding these `FutureWarnings` behind `verbose` would be better.

Comment (thomwolf) [reply src/transformers/tokenization_utils_base.py]: yes

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -1960,6 +1960,220 @@ def pad(
...
+                truncation_strategy=truncation_strategy,
+                stride=stride,
+            )
+            if return_overflowing_tokens:
+                encoded_inputs["overflowing_tokens"] = overflowing_tokens
+                encoded_inputs["num_truncated_tokens"] = total_len - max_length
+
+        # Add special tokens
+        if add_special_tokens:
+            sequence = self.build_inputs_with_special_tokens(ids, pair_ids)
```
Comment (n1t0) [inline src/transformers/tokenization_utils_base.py:2064]: I think the only problem is that this method and the one used below are not overridden on the fast tokenizers, so it doesn't add the special tokens.

Comment (thomwolf) [reply src/transformers/tokenization_utils_base.py:2064]: This is now fixed actually (I was wrong when I told you it was broken).

**Review by thomwolf — APPROVED** [review]
Great!

**Pull request merged**